### ISSUE TITLE
Slow Firing / Slow Turning Turrets Miss Moving Targets.

### ISSUE THREAD

**Issue by TheKurgan** [PapaJoesSoup/BDArmory#497]
ISSUE TYPE
 - Bug Report   (Something is... not working well)

KSP and BDA Version

KSP 13.1, 1.4.1, 1.4.2

BDA version (all) First noticed in January

OS / ENVIRONMENT
Win 10

SUMMARY
Slow rate of fire Turrets (tank turrets like the Mi Abrams), miss moving targets:

Machine gun turrets seem fine, but the slower rate of fire and slower turning turrets, are firing before they are on target.... then they pause/freeze while reloading, then when the turret goes to snap back to the target, it fires again BEFORE it's on target and misses. rinse repeat.

The other scenario is when the turret snaps to target, pauses/freezes for a split second then fires and the target has already moved out of the way. 

Again, this would not be noticed using fast firing turrets, OR in situations against against stationary targets or between two tanks coming directly towards each other.

Once in a while... just by luck, (fluke timing) it fires right after the turret is "on target" and I get a hit.

##### STEPS TO REPRODUCE

Use the M1 Abrams turret (stationary) in guard mode.
Place a WM enabled craft about 1km to 1.5km away moving perpendicular to the target at 20+ ms
See if Guard mode can hit it.

**Comment by PapaJoesSoup** [comment]
We appreciate the steps to reproduce, but we will still need a log file.

**Comment by TheKurgan** [comment]
This has been noticed by me, Spanner, and a few other members of the BDAc team. It isn't local to my install. It's a confirmed behavior that we have been discussing on Discord.
I can still make a log of it on a clean install later.

**Comment by TheKurgan** [comment]
Interesting, can you elaborate for me? Are you setting it for the reload of the weapon or something?

**Comment by Deathpuff12** [comment]
Sorry for the late reply. I just set the "Firing Burst Length" (wrong on the name, oops) to an arbitrary value; it's set to 5 on all my craft. Once you've done that, the turret will continuously track the target and shoot. Also, not sure if it makes a difference, but I have "Firing Interval" set to 3.
This doesn't, however, fix the initial shot. But it will keep the turret aiming at the target. I've even tested it by circling around an enemy vehicle, and the turret tracked and fired just fine.

**Comment by RichTheWolf257** [comment]
That sounds more like a balance feature than a bug LOL! A slow moving, slow firing turret WILL tend to miss moving targets XD

**Comment by TheKurgan** [comment]
Not really, tell that to a guy firing the gun on a modern main battle tank.
I'm not talking about a moving target that is doing 100mph, I'm talking about another tank moving perpendicular to the turret that is firing... doing maybe 25ms
This also applies to Turrets like my Turreted Rail gun, that fires a projectile at 5000m/s
The issue is, the slow rate of fire turrets are not tracking the target during reload, and end up firing at the wrong moment.

**Comment by SpannerMonkey** [comment]
Update
BDA BDArmory.1.2.1.1_05192018
KSP 1.4.3

The issue remains 
Tank turrets and other slow firing targets are firing at where the target was not where it will be.
At 1km it's a 3-10 hit rate and longer ranges the chances of a hit are minimal.  

Oddly however it only seems to be very noticeable on craft that are landed on terrain,  craft splashed do not seem to have such a pronounced problem ( could just be marine targets being much larger) 

I still feel that is related to the camera and it's interaction with the AI, the point fired at is always the location of camera focus,  which always lags behind a moving craft by several meters.  It can be clearly seen in the log that the target is indeed set to camera focus , and not the moving craft or it's COM

**Comment by XOC2008** [comment]
From Discord:
[10:40 PM] Spanner: what i want it to do 
Turret on guard mode
Target at 2km moving left to right
ANY turret should be able to track the moving targets com.
Any turret should be able to lead enough to hit a moving target  regardless of shell mass and velocity 
The com target on a ground engagement should take precedence over any other form.
If turret has LOS to a target it should run on pure raycast from the fire transform.
Only if target is obscured should it use another form of targetting , if available,
In preference   Fire transform,  Radar,  and as we havent ot a kerb with binos, I'll have to put  cam

**Comment by SpannerMonkey** [comment]
List of basic requirements for turret operations in BDA 
```
Turret on guard mode
Target at 2km moving left to right
ANY turret should be able to track the moving targets com.
Any turret should be able to lead enough to hit a moving target  regardless of shell mass and velocity 
The com target on a ground engagement should take precedence over any other form.
If turret has LOS to a target it should run on pure raycast from the fire transform.
Only if target is obscured should it use another form of targetting , if available,
In preference   Fire transform,  Radar,  and finally if all else fails use the cam system 
```

**Comment by Gedas-S** [comment]
Should be fixed in dev (#570), needs confirmation though (aka works on my box, at least). :)
